tools/update_ruby_gem_packages.rb

```diff
@@ -5,55 +5,77 @@
     max_queue: 0, # unbounded work queue
     fallback_policy: :caller_runs
   )
+
+  # Currently, the only packages that use the ruby buildsystem are ruby_* packages, so this finds all the packages we need to check.
   relevant_gem_packages = Dir['packages/ruby_*.rb']
+
+  # Get the total number of files to check, and then the length of that number, so status updates can be formatted.
   total_files_to_check = relevant_gem_packages.length
   numlength = total_files_to_check.to_s.length
   updateable_packages = {}
+  packages_without_gem_versions = []
   relevant_gem_packages.each_with_index do |package, index|
     pool.post do
-      untested_package_name = package.gsub(%r{^packages/ruby_}, '').gsub(/.rb$/, '')
-      gem_test = gems.grep(/#{"^#{untested_package_name}\\s.*$"}/).last.blank? ? gems.grep(/#{"^\(#{passed_name.gsub(/^ruby_/, '').gsub('_', ')*.(')}\\s\).*$"}/).last : gems.grep(/#{"^#{untested_package_name}\\s.*$"}/).last
-      abort "Cannot find #{passed_name} gem to install.".lightred if gem_test.blank?
-      gem_test_name = gem_test.split.first
-      gem_test_versions = gem_test.split[1].split(',')
-      # Any version with a letter is considered a prerelease as per
-      # https://github.com/rubygems/rubygems/blob/b5798efd348935634d4e0e2b846d4f455582db48/lib/rubygems/version.rb#L305
-      gem_test_versions.delete_if { |i| i.match?(/[a-zA-Z]/) }
-      gem_test_version = gem_test_versions.max
-      gem_name = gem_test_name.blank? ? Gem::SpecFetcher.fetcher.suggest_gems_from_name(untested_package_name).first : gem_test_name
-      gem_version = gem_test_name.blank? ? Gem.latest_version_for(untested_package_name).to_s : gem_test_version
-      next package if gem_version.blank?
-
-      relevant_gem_packages.delete(package)
+      pkg = Package.load_package(package)
+      gem_name = pkg.name.sub('ruby_', '')
+      # We replace all dashes with underscores in our initial package names, but some gems actually use underscores, so we need special cases.
+      # This list was created by looking at what packages were listed as not having updates in rubygems, and then looking up the upstream name for them.
+      if %w[connection_pool error_highlight mini_mime multi_xml mutex_m power_assert regexp_parser repl_type_completor ruby2_keywords syntax_suggest].include?(gem_name)
+        # These gems used underscores originally, so don't replace anything
+      elsif gem_name == 'language_server_protocol'
+        # These gems have an underscore then a dash, but there's only one, so we hardcode the logic for now.
+        gem_name = 'language_server-protocol'
+      elsif gem_name == 'unicode_display_width'
+        # These gems have a dash then an underscore, but there's only one, so we hardcode the logic for now.
+        gem_name = 'unicode-display_width'
+      else
+        # In the common case, the gem name used only dashes, which we all replaced with underscores.
+        gem_name.gsub!('_', '-')
+      end
+
```

Comment:
I don't think that whitelisting all gems with variant usage of dashes & underscores is resilient to a new package having the same issue?

What's wrong with just having a code path that checks for all possibilities if a dash or underscore may be in the gem name, like before?

Reply:
The previous codepath didn't work:
```
Done checking for updates to 133 ruby gems.
Updated versions were could not be found for: concurrent-ruby gem-compiler io-console io-nonblock io-wait language-server-protocol net-ftp net-http net-imap net-pop net-protocol net-smtp net-telnet open-uri pry-byebug readline-ext resolv-replace rubocop-ast rubocop-chromebrew ruby-libversion ruby-progressbar rubygems-update test-unit unicode-display-width unicode-emoji yaml-lint
```

Reply:
Actually, that raises the question of whether the logic for installing these gems works properly, given that its the one I replaced here.

Reply:
Would it make more sense to just have a property in gem packages denoting their upstream name where it differs from a simple `.gsub('_','-')`? (Obviously that'd be something for a different PR)

Reply:
Maybe? Or maybe we should figure out a way to allow dashes and underscores in our package names?

Reply:
That would also be a good solution-- something for a second PR, perhaps?